sqlalchemy-stubs/sql/elements.pyi

```diff
@@ -47,300 +61,388 @@ class ColumnElement(operators.ColumnOperators, ClauseElement):
+    def compare(self, other: Any, **kw: Any) -> bool: ...
+
+_BCL = TypeVar('_BCL', bound=BooleanClauseList)
+
+class BooleanClauseList(ClauseList, ColumnElement[Boolean]):
     __visit_name__: str = ...
-    def __init__(self, *arg, **kw) -> None: ...
     @classmethod
-    def and_(cls, *clauses) -> BooleanClauseList: ...
+    def and_(cls, *clauses: ClauseElement) -> BooleanClauseList: ...
     @classmethod
-    def or_(cls, *clauses) -> BooleanClauseList: ...
-    def self_group(self, against: Optional[Any] = ...): ...
+    def or_(cls, *clauses: ClauseElement) -> BooleanClauseList: ...
+    def self_group(self: _BCL, against: Optional[Any] = ...) -> Union[_BCL, Grouping[_TE]]: ...
 
 and_ = BooleanClauseList.and_
 or_ = BooleanClauseList.or_
 
-class Tuple(ClauseList, ColumnElement):
-    type: Any = ...
-    def __init__(self, *clauses, **kw) -> None: ...
+class Tuple(ClauseList, ColumnElement[_TE]):
+    type: _TE = ...
+    @overload
+    def __init__(self, *clauses: ColumnElement[Any], type_: Type[_TE], **kw: Any) -> None: ...
+    @overload
+    def __init__(self, *clauses: ColumnElement[Any], type_: _TE, **kw: Any) -> None: ...
+    @overload
+    def __init__(self, clause: ColumnElement[_TE], *clauses: ColumnElement[Any], **kw: Any) -> None: ...
 
-class Case(ColumnElement):
+class Case(ColumnElement[TypeEngine[_T]], Generic[_T]):
     __visit_name__: str = ...
     value: Any = ...
-    type: Any = ...
+    type: TypeEngine[_T] = ...
     whens: Any = ...
     else_: Any = ...
-    def __init__(self, whens, value: Optional[Any] = ..., else_: Optional[Any] = ...) -> None: ...
-    def get_children(self, **kwargs): ...
+    @overload
+    def __init__(self, whens: Iterable[_TupleType[ClauseElement, _T]], else_: Optional[_T] = ...) -> None: ...
+    @overload
+    def __init__(self, whens: Mapping[Any, _T], value: ColumnElement[Any] = ..., else_: Optional[_T] = ...) -> None: ...
+    def get_children(self, **kwargs: Any) -> Any: ...
 
-def literal_column(text, type_: Optional[Any] = ...): ...
+@overload
+def literal_column(text: str, type_: Optional[Type[_TE]] = ...) -> ColumnClause[_TE]: ...
+@overload
+def literal_column(text: str, type_: Optional[_TE] = ...) -> ColumnClause[_TE]: ...
 
-class Cast(ColumnElement):
-    __visit_name__: str = ...
-    type: Any = ...
-    clause: Any = ...
-    typeclause: Any = ...
-    def __init__(self, expression, type_) -> None: ...
-    def get_children(self, **kwargs): ...
+_OTE = TypeVar('_OTE', bound=TypeEngine[Any])
 
-class TypeCoerce(ColumnElement):
+class Cast(ColumnElement[_TE], Generic[_TE, _OTE]):
     __visit_name__: str = ...
-    type: Any = ...
-    clause: Any = ...
-    def __init__(self, expression, type_) -> None: ...
-    def get_children(self, **kwargs): ...
+    type: _TE = ...
+    clause: ClauseElement = ...
+    typeclause: TypeClause[_TE] = ...
+    @overload
+    def __init__(self, expression: Union[str, ColumnElement[_OTE]], type_: Type[_TE]) -> None: ...
```

Comment:
There are many errors on calls to cast, I think it accepts `Iterable[str]` and `Iterable[unicode]` as well. Also I am not sure adding `_OTE` is a good idea, this type does not appear anywhere except constructor, on giving a string there (common case) will generate `Need type annotation for variable`.

Reply:
Yeah, after further thinking about it, `expression` should just be `Any` and the class then becomes much simpler. I've updated the code.